DPJTH-4 Add unit tests to HSM sample

datalogics-robl opened the pull request:
#### Changes in this pull request

Add several unit tests to the HSM sample 
#### Fulfills  [DPJTH-4](https://jira.datalogics.com/browse/DPJTH-4)
#### Checklist for approving this pull request

(PR creator amend this with more conditions if necessary)
- [x] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [x] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [x] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [x] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [x] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)
#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:
- [x] _&lt;add your name here with an **@**>_

datalogics-robl committed:
Make sure the HsmManager can only be used if connected

By treating hsmManager as a member variable with state, we opened the possiblity
of trying to use it in an improper state: a direct call to
signExistingSignatureFields() would crash, since the hsmManager had never been
connected to the device.

We now require clients to supply their own, connected HsmManager.

datalogics-robl committed:
Test that an HsmManager must be connected to be used

This includes a MockHsmManager that does not implement any methods: relevant
methods are overridden by the individual tests.

datalogics-robl committed:
Throw IllegalStateExcpetion if HsmManager is not logged in

datalogics-robl committed:
Verify exception and log record for unconnected HSM test

datalogics-robl committed:
Rename unit test to 'UnconnectedHsmTest'

This unit test will ONLY test the HsmSignDocument sample with a mocked
unconnected HSM device.

datalogics-robl committed:
Refactor UnconnectedHsmTest into individual unit tests

This now tests three behaviors separately:
- An unconnected HSM causes an exception to be thrown
- An unconnected HSM generates a log message
- An unconnected HSM causes no output to be generated

datalogics-robl committed:
Factor the unconnected HsmManager into an abstract class

This promotes code reuse by other tests: the AbstractHsmManager allows tests
to override only the HsmManager APIs they require.

datalogics-robl committed:
Copy certificate and key from regular signing sample

datalogics-robl committed:
Add a mocked HSM Manager test

This mocks the HsmManager. Note that it currently does not pass, nor are the
actual test calls in their finished form.

datalogics-robl committed:
Add a mock security provider and mock signer for testing

The MockSigner wraps a java.security.Signature object and uses it for
the work of signing. The MockProvider makes it available to the JCA
system. This allows us to emulate HSM hardware in software.

datalogics-robl committed:
Modify MockedHsmTest to use MockProvider

This test now passes, and produces a signed PDF using the mocked HSM
provider.

datalogics-robl committed:
Move the AbstractHsmManager to the mock package

datalogics-robl committed:
Add unit test to validate the signature produced by the sample

This replicates code from the SignDocumentTest in the main samples.

datalogics-robl committed:
Remove exception catch blocks from MockedHsmTest

These were copied from the UnconnectedHsmTest; in this case, exceptions
are *not* expected, and constitute a test failure if obtained.

datalogics-robl committed:
Add HsmManagerFactoryTest

Mocks the LunaSlotManager and LunaProvider to provide a fully
functioning LunaHsmManager.

datalogics-robl committed:
Add LunaHsmManagerTest

datalogics-robl committed:
Add a MockKeyStore to our MockProvider

This makes a KeyStore available under our MockProvider. This
implementation contains nothing but stub methods.

datalogics-robl committed:
Add connection state test for LunaHsmManager

Make sure we properly track the current state for a login/logout cycle.

datalogics-robl committed:
Refactor LunaHsmManager into a global variable

datalogics-robl committed:
Remove redundant connection state check in tests

This was copied from the sample code, but makes little sense in a
unit test.

datalogics-robl committed:
Add test for correct provider name

datalogics-robl committed:
Separate out slot handling into separate tests

The slot parameter (aka 'token') is a special issue with the Luna HSM
and deserves separate testing. This adds distinct unit tests for
logging in with this parameter, and makes sure our mock LunaSlotManager
behaves like the real device.

datalogics-robl committed:
Update LunaHsmManager to handle bad slot names

The LunaSlotManager treats a bad slot name differently from a bad
password: it does not throw an exception but quietly fails to log in.
This was revealed in unit testing.

To handle this behavior, we make two changes:
- Only call loadKeyStore() if we are definitely logged in
- Throw an IllegalArgumentException if are not logged in

Note that we also eliminate the 'logged in' check in loadKeyStore(),
since it is now a private method that should only be called when we
are in the correct state.

datalogics-robl committed:
Add ability to return a private key from the MockKeyStore

datalogics-robl committed:
Add canRetrieveKey() test to LunaHsmManagerTest

datalogics-devauto commented:
Build succeeded.

jrowden8 commented on src/main/java/com/datalogics/pdf/security/LunaHsmManager.java:
```diff
@@ -71,8 +71,10 @@ public void hsmLogin(final HsmLoginParameters parms) {
             state = ConnectionState.CONNECTED;
+            loadKeyStore();
+        } else {
+            throw new IllegalArgumentException("Could not log into the Luna HSM");
```
👍

jrowden8 replied:
Should the Exception thrown at line 76 be a IllegalArgumentException or a IllegalStateException seeing that the manager not connecting  could be for reasons other than the login parameters passed to this method

datalogics-robl replied:
That's a tricky one, because the specific behavior I was trying to prevent with this _is_ due to a bad argument: a bad token/slot name. However, that's not the _only_ reason the log in might fail, and there's not an obvious way to even distinguish among a bad slot name and other causes, let alone report it.

jrowden8 commented on src/test/java/com/datalogics/pdf/security/LunaHsmManagerTest.java:
```diff
@@ -0,0 +1,211 @@
+
+        lunaHsmManager.hsmLogin(new LunaHsmLoginParameters(BAD_PASSWORD));
+
+        assertEquals("LunaHsmManager login should be unsuccessful", ConnectionState.CONNECTED,
```
Is the assertEquals at line 77 really needed as the expected IllegalArgumentException will be thrown at line 75 and this test will pass?

jrowden8 commented on src/test/java/com/datalogics/pdf/security/LunaHsmManagerTest.java:
```diff
@@ -0,0 +1,211 @@
+/**
+ * Unit tests for the LunaHsmManager.
+ */
+public class LunaHsmManagerTest {
```
Lots of really good tests of the LunaHsmManager here. Nice!

jrowden8 commented on src/test/java/com/datalogics/pdf/security/LunaHsmManagerTest.java:
```diff
@@ -0,0 +1,211 @@
+
+        lunaHsmManager.hsmLogin(new LunaHsmLoginParameters(BAD_SLOT_NAME, GOOD_PASSWORD));
+
+        assertEquals("LunaHsmManager login with bad slot name should be unsuccessful", ConnectionState.CONNECTED,
```
Similar to line 77 is the assertEquals() at line 101 really necessary?

jrowden8 commented on src/test/java/com/datalogics/pdf/hsm/samples/MockedHsmTest.java:
```diff
@@ -0,0 +1,227 @@
+        try {
+            HsmSignDocument.signExistingSignatureFields(connectedHsmManager, inputUrl, outputUrl);
+        } catch (final IllegalStateException e) {
+            // Expected exception
```
Is this try-catch block really needed or is there more work to be done here?

datalogics-kam commented:
I'm wondering if we should have more distinction between how we name Mocks, Fakes, and Stubs. For discussion: http://stackoverflow.com/a/346440

jrowden8 commented:
This is looking really good so far. The overall functionality of the HSM sample, HsmManagerFactory, and the LunaHsmManager appear to have tests covering them.

datalogics-kam commented:
The integration tests don't run, of course, because I don't have access to hardware. They should always be run in CI. The question is how do we solve this problem:
1. Have the ability to request mocks for the integration tests maybe by using a profile? (this is probably a really good idea)
2. Somehow make the HSM available to all and sundry users?
3. Get a Jenkins slave on the machine that has access to the HSM? (Is this possible? Do we need a server?)

datalogics-robl commented:
As mentioned in the pull request message, I see two things that could still be tested:
- _LunaHsmManager.getCertificateChain()_: Verify that `LunaHsmManager` interacts appropriately with a certificate provider. This is worth testing, but I'm doubtful I can complete it in time for the sprint end.
- _LunaHsmLoginParameters_: Is just a property holder class; there is not currently any logic in it to test.

One question I'm posing with this PR is whether we want to get those tests in before saying we're satisfied that the card is complete.

datalogics-robl commented:
> The integration tests don't run, of course, because I don't have access to hardware. They should always be run in CI. The question is how do we solve this problem...

I advocate for option 1 in @datalogics-kam's list, but again, that will take time to implement.

datalogics-kam commented:
One way to figure out if we've done enough testing is to run a coverage test. I loaded EclEmma into my Eclipse, and ran the unit tests only (because I can't run the integration tests), and here's sort of a snapshot of what the coverage looks like.

![screen shot 2016-06-01 at 2 52 55 pm](https://cloud.githubusercontent.com/assets/89971/15723586/c1968b90-2808-11e6-9a47-83b13cb09a74.png)

It looks like the security package needs more testing.

Also, FontUtils.java has no coverage at all; it also has no references. I'm wondering if we shouldn't drop it from the project altogether.

datalogics-kam commented:
> I advocate for option 1 in @datalogics-kam's list, but again, that will take time to implement.

Option 1 is a great solution, and I would also like to see us running Jenkins on a machine that has HSM access.

datalogics-kam commented:
Bear in mind that some of the coverages are low, like `HSMSignDocument` shows 58%, but that's not too bad, because the part that is covered is the juicy logic, and the part that isn't covered is `main()`, so it helps to open the source in the editor and look over the line-by-line coverage.

datalogics-robl commented:
??? How can `HsmManager` have 81.2% coverage and `LunaHsmManager` have 0.0%? `HsmManager` isn't even _code_: it's an interface!

datalogics-kam commented:
> ??? How can HsmManager have 81.2% coverage and LunaHsmManager have 0.0%? HsmManager isn't even code: it's an interface!

It successfully initialized its enum:

![screen shot 2016-06-01 at 2 58 59 pm](https://cloud.githubusercontent.com/assets/89971/15723711/6571e20a-2809-11e6-9c02-61fea17f363d.png)

The secret is that declaring enumeration values causes the construction of the objects in a static initializer, so that's code.

datalogics-robl committed:
Remove unnecessary asserts

datalogics-robl committed:
Remove unnecessary try/catch block

datalogics-robl committed:
Move AllUnitTests and AllIntegrationTests to com.datalogics.pdf

Since we have some tests in the com.datalogics.pdf.security package,
they were not being run by AllUnitTests. The relocation catches *all*
tests.

datalogics-devauto commented:
Build succeeded.

datalogics-robl committed:
Rename 'Mock' classes to 'Fake' classes, plus package name

'Mocks' usually include expectation tests for how they are called; these
classes simply provide sparse-but-functioning implementations of
objects normally provided by the HSM.

datalogics-devauto commented:
Build succeeded.

datalogics-robl committed:
Explicitly check for supported slot name

We can query the LunaSlotManager for whether a slot name is supported.
Includes test to confirm.

datalogics-robl committed:
Throw IllegalStateException if we are not logged in

datalogics-devauto commented:
Build succeeded.

jrowden8 commented on src/test/java/com/datalogics/pdf/hsm/samples/MockedHsmTest.java:
```diff
@@ -0,0 +1,223 @@
+/**
+ * Unit tests simulating a connected HSM device.
+ */
+public class MockedHsmTest extends SampleTest {
```
Did we ever decide on a naming convention for "mocked" classes that do not extend MockUp?

As @datalogics-kam  mentioned I can see where this could be confusing being named MockedHsmTest when there are no "@mock"-ed tests listed here in this class.

datalogics-robl replied:
I did rename several helper classes from `Mock*` to `Fake*`, but I didn't think about this issue. It's true that this class has no "@mock"-ed tests.

I struggled with what to name this test (to distinguish it from `UnconnectedHsmTest`: `ConnectedHsmTest` would be misleading since it doesn't connect to a _real_ HSM. `FakeLunaHsmTest` perhaps?

datalogics-robl committed:
Rename 'MockedHsmTest' to 'FakeConnectedHsmTest'

'Mocked' classes includes some instrumentation to verify they are being
called correctly; this is more accurately described as a 'fake'.

datalogics-devauto commented:
Build succeeded.

jrowden8 commented:
Looks good to me and builds successfully on my machine. Some stellar work here @datalogics-robl!
